consensus/blake3pow/consensus.go

```diff
@@ -331,48 +334,41 @@ func (blake3pow *Blake3pow) CalcDifficulty(chain consensus.ChainHeaderReader, pa
 		log.Error("Cannot CalcDifficulty for", "context", nodeCtx)
 		return nil
 	}
-	// algorithm:
-	// diff = (parent_diff +
-	//         (parent_diff / 2048 * max((2 if len(parent.uncles) else 1) - ((timestamp - parent.timestamp) // 9), -99))
-	//        ) + 2^(periodCount - 2)
 
-	time := parent.Time()
+	///// Algorithm:
+	///// e = (DurationLimit - (parent.Time() - parentOfParent.Time())) * parent.Difficulty()
+	///// k = Floor(BinaryLog(parent.Difficulty()))/(DurationLimit*DifficultyAdjustmentFactor*AdjustmentPeriod)
+	///// Difficulty = Max(parent.Difficulty() + e * k, MinimumDifficulty)
 
 	if parent.Hash() == chain.Config().GenesisHash {
 		return parent.Difficulty()
 	}
 
 	parentOfParent := chain.GetHeaderByHash(parent.ParentHash())
+	if parentOfParent.Hash() == chain.Config().GenesisHash {
+		return parent.Difficulty()
+	}
 
+	time := parent.Time()
 	bigTime := new(big.Int).SetUint64(time)
 	bigParentTime := new(big.Int).SetUint64(parentOfParent.Time())
 
 	// holds intermediate values to make the algo easier to read & audit
 	x := new(big.Int)
-	y := new(big.Int)
-
-	// (2 if len(parent_uncles) else 1) - (block_timestamp - parent_timestamp) // duration_limit
 	x.Sub(bigTime, bigParentTime)
+	x.Sub(blake3pow.config.DurationLimit, x)
+	x.Mul(x, parent.Difficulty())
+	k, _ := mathutil.BinaryLog(new(big.Int).Set(parent.Difficulty()), 64)
```

Comment:
Is it sufficient to have k just be an integer?

Reply:
Yes, it is, it is us trying to represent the bits using the difficutly, if you introduce the mantissa, we will have to deal with Floats, so for now, we are using the characterstic value only